pkg/provider/cloud/vsphere/provider.go

```diff
@@ -48,109 +39,44 @@ const (
-		kruntime.HandleError(fmt.Errorf("vSphere client failed to logout: %w", err))
-	}
-}
+func (v *VSphere) reconcileCluster(ctx context.Context, cluster *kubermaticv1.Cluster, update provider.ClusterUpdater, force bool) (*kubermaticv1.Cluster, error) {
+	logger := v.log.With("cluster", cluster.Name)
 
-func newSession(ctx context.Context, dc *kubermaticv1.DatacenterSpecVSphere, username, password string, caBundle *x509.CertPool) (*Session, error) {
-	u, err := url.Parse(fmt.Sprintf("%s/sdk", dc.Endpoint))
+	username, password, err := getCredentialsForCluster(cluster.Spec.Cloud, v.secretKeySelector, v.dc)
 	if err != nil {
 		return nil, err
 	}
 
-	// creating the govmoni Client in roundabout way because we need to set the proper CA bundle: reference https://github.com/vmware/govmomi/issues/1200
-	soapClient := soap.NewClient(u, dc.AllowInsecure)
-	// set our CA bundle
-	soapClient.DefaultTransport().TLSClientConfig.RootCAs = caBundle
-
-	vim25Client, err := vim25.NewClient(ctx, soapClient)
-	if err != nil {
-		return nil, err
-	}
-
-	client := &govmomi.Client{
-		Client:         vim25Client,
-		SessionManager: session.NewManager(vim25Client),
-	}
-
-	user := url.UserPassword(username, password)
-	if dc.InfraManagementUser != nil {
-		user = url.UserPassword(dc.InfraManagementUser.Username, dc.InfraManagementUser.Password)
-	}
-
-	if err = client.Login(ctx, user); err != nil {
-		return nil, fmt.Errorf("failed to login: %w", err)
-	}
-
-	finder := find.NewFinder(client.Client, true)
-	datacenter, err := finder.Datacenter(ctx, dc.Datacenter)
-	if err != nil {
-		return nil, fmt.Errorf("failed to get vSphere datacenter %q: %w", dc.Datacenter, err)
-	}
-	finder.SetDatacenter(datacenter)
-
-	return &Session{
-		Datacenter: datacenter,
-		Finder:     finder,
-		Client:     client,
-	}, nil
-}
-
-// getVMRootPath is a helper func to get the root path for VM's
-// We extracted it because we use it in several places.
-func getVMRootPath(dc *kubermaticv1.DatacenterSpecVSphere) string {
-	// Each datacenter root directory for VM's is: ${DATACENTER_NAME}/vm
-	rootPath := path.Join("/", dc.Datacenter, "vm")
-	// We offer a different root path though in case people would like to store all Kubermatic VM's below a certain directory
-	if dc.RootPath != "" {
-		rootPath = path.Clean(dc.RootPath)
-	}
-	return rootPath
-}
-
-// InitializeCloudProvider initializes the vsphere cloud provider by setting up vm folders for the cluster.
-func (v *Provider) InitializeCloudProvider(ctx context.Context, cluster *kubermaticv1.Cluster, update provider.ClusterUpdater) (*kubermaticv1.Cluster, error) {
-	username, password, err := GetCredentialsForCluster(cluster.Spec.Cloud, v.secretKeySelector, v.dc)
-	if err != nil {
-		return nil, err
-	}
 	rootPath := getVMRootPath(v.dc)
-	if cluster.Spec.Cloud.VSphere.Folder == "" {
+	if force || cluster.Spec.Cloud.VSphere.Folder == "" {
+		logger.Infow("reconciling vsphere folder", "folder", cluster.Spec.Cloud.VSphere.Folder)
```

Comment:
I cannot comment on the lines below this, but this will become a problem. The code within this if condition always adds the finalizer to clean up the vSphere folder, which in turn means that `CleanUpCloudProvider` will delete it. However, KKP should only delete resources it created.

Reply:
The problem is, resources are classified as two categories, resources create by KKP as defaulting or resources that are created beforehand and referenced in. Now, this is a bit problematic, since there are no claims on those resources, as we don't identify what has been created by KKP and what has been created by the users.

Reply:
However, the folder which is created by KKP and it follows the cluster name. Maybe we can add more semantics into this, so KKP cleans up only what it has created.